File: src/turbopuffer_fs/live.py
"""Thin live wrappers over plans, runtime, and finalizers."""

from __future__ import annotations

from .fs import (
    cat_plan,
    find_plan,
    grep_plan,
    head_plan,
    ls_plan,
    mkdir_plan,
    put_bytes_plan,
    put_text_plan,
    read_bytes_plan,
    read_text_plan,
    rm_plan,
    stat_plan,
    tail_plan,
)
from .runtime import run


MOUNT_SUFFIX = "__fs"


def mount_namespace(mount: str) -> str:
    if not isinstance(mount, str):
        raise TypeError("mount must be a string")
    if mount == "":
        raise ValueError("mount must not be empty")
    if "/" in mount:
        raise ValueError(f"mount must not contain '/': {mount!r}")
    return f"{mount}{MOUNT_SUFFIX}"


def make_client(*, api_key: str | None = None, region: str | None = None, base_url: str | None = None, **kwargs):
    import turbopuffer

    values = dict(kwargs)
    if api_key is not None:
        values["api_key"] = api_key
    if region is not None:
        values["region"] = region
    if base_url is not None:
        values["base_url"] = base_url
    return turbopuffer.Turbopuffer(**values)


def _mounts_plan(*, suffix: str = MOUNT_SUFFIX) -> dict[str, object]:
    return {
        "namespace": "",
        "steps": [{"kind": "namespaces", "name": "namespaces", "payload": {}}],
        "finalize": "mounts",
        "context": {"suffix": suffix},
    }


def list_mounts(client, *, suffix: str = MOUNT_SUFFIX) -> list[str]:
    return run(client, _mounts_plan(suffix=suffix))


def stat(client, mount: str, path: str):
    return run(client, stat_plan(mount_namespace(mount), path))


def ls(client, mount: str, path: str = "/", *, limit: int | None = None):
    return run(client, ls_plan(mount_namespace(mount), path, limit=limit))


def find(
    client,
    mount: str,
    root: str = "/",
    *,
    glob: str | None = None,
    kind: str | None = None,
    ignore_case: bool = False,
    limit: int | None = None,
):
    return run(
        client,
        find_plan(
            mount_namespace(mount),
            root,
            glob=glob,
            kind=kind,
            ignore_case=ignore_case,
            limit=limit,
        ),
    )


def cat(client, mount: str, path: str) -> str:
    return run(client, cat_plan(mount_namespace(mount), path))


def head(client, mount: str, path: str, n: int = 10) -> list[str]:
    return run(client, head_plan(mount_namespace(mount), path, n=n))


def tail(client, mount: str, path: str, n: int = 10) -> list[str]:
    return run(client, tail_plan(mount_namespace(mount), path, n=n))


def grep(
    client,
    mount: str,
    root: str,
    pattern: str,
    *,
    ignore_case: bool = False,
    glob: str | None = None,
    limit: int | None = None,
):
    return run(
        client,
        grep_plan(
            mount_namespace(mount),
            root,
            pattern,
            ignore_case=ignore_case,
            glob=glob,
            limit=limit,
        ),
    )


def read_text(client, mount: str, path: str) -> str:
    return run(client, read_text_plan(mount_namespace(mount), path))


def read_bytes(client, mount: str, path: str) -> bytes:
    return run(client, read_bytes_plan(mount_namespace(mount), path))


def mkdir(client, mount: str, path: str):
    return run(client, mkdir_plan(mount_namespace(mount), path))


def put_text(client, mount: str, path: str, text: str, *, mime: str | None = None):
    return run(client, put_text_plan(mount_namespace(mount), path, text, mime=mime))


def put_bytes(client, mount: str, path: str, data: bytes, *, mime: str | None = None):
    return run(client, put_bytes_plan(mount_namespace(mount), path, data, mime=mime))


def rm(client, mount: str, path: str, *, recursive: bool = False):
    return run(client, rm_plan(mount_namespace(mount), path, recursive=recursive))

File: packages/turbopuffer-fs/src/live.ts
import Turbopuffer from "@turbopuffer/turbopuffer";

import {
  catPlan,
  findPlan,
  grepPlan,
  headPlan,
  lsPlan,
  mkdirPlan,
  putBytesPlan,
  putTextPlan,
  readBytesPlan,
  readTextPlan,
  rmPlan,
  statPlan,
  tailPlan,
} from "./plans.js";
import { run } from "./runtime.js";
export { ingestDirectory } from "./ingest.js";
import type { GrepOptions } from "./types.js";

export const MOUNT_SUFFIX = "__fs";

export function mountNamespace(mount: string): string {
  if (mount.length === 0) {
    throw new Error("mount must not be empty");
  }
  if (mount.includes("/")) {
    throw new Error(`mount must not contain '/': ${mount}`);
  }
  return `${mount}${MOUNT_SUFFIX}`;
}

export function makeClient(options: {
  apiKey?: string;
  region?: string;
  baseURL?: string;
}) {
  return new Turbopuffer({
    apiKey: options.apiKey,
    region: options.region,
    baseURL: options.baseURL,
  });
}

export async function listMounts(client: Turbopuffer, suffix = MOUNT_SUFFIX): Promise<string[]> {
  const response = await client.namespaces();
  const namespaces: Array<{ id: string }> = [];
  for await (const namespace of response) {
    namespaces.push({ id: namespace.id });
  }
  const mounts = namespaces
    .map((namespace) => namespace.id)
    .filter((id: string) => id.endsWith(suffix))
    .map((id: string) => id.slice(0, -suffix.length));
  mounts.sort();
  return mounts;
}

export async function runPlan(client: Turbopuffer, plan: import("./types.js").Plan): Promise<unknown> {
  return run(client, plan);
}

export function stat(client: Turbopuffer, mount: string, path: string) {
  return runPlan(client, statPlan(mountNamespace(mount), path));
}

export function ls(client: Turbopuffer, mount: string, path = "/", limit?: number) {
  return runPlan(client, lsPlan(mountNamespace(mount), path, limit));
}

export function find(
  client: Turbopuffer,
  mount: string,
  root = "/",
  options: { glob?: string; kind?: "file" | "dir"; ignoreCase?: boolean; limit?: number } = {},
) {
  return runPlan(
    client,
    findPlan(mountNamespace(mount), root, {
      glob: options.glob,
      kind: options.kind,
      ignoreCase: options.ignoreCase,
      limit: options.limit,
    }),
  );
}

export function cat(client: Turbopuffer, mount: string, path: string) {
  return runPlan(client, catPlan(mountNamespace(mount), path));
}

export function head(client: Turbopuffer, mount: string, path: string, n = 10) {
  return runPlan(client, headPlan(mountNamespace(mount), path, n));
}

export function tail(client: Turbopuffer, mount: string, path: string, n = 10) {
  return runPlan(client, tailPlan(mountNamespace(mount), path, n));
}

export function grep(
  client: Turbopuffer,
  mount: string,
  root: string,
  pattern: string,
  options: GrepOptions = {},
) {
  return runPlan(
    client,
    grepPlan(mountNamespace(mount), root, pattern, {
      ignoreCase: options.ignoreCase,
      glob: options.glob,
      limit: options.limit,
      mode: options.mode,
      multiline: options.multiline,
      dotAll: options.dotAll,
      lastAsPrefix: options.lastAsPrefix,
    }),
  );
}

export function search(
  client: Turbopuffer,
  mount: string,
  root: string,
  pattern: string,
  options: GrepOptions = {},
) {
  return grep(client, mount, root, pattern, options);
}

export function readText(client: Turbopuffer, mount: string, path: string) {
  return runPlan(client, readTextPlan(mountNamespace(mount), path));
}

export function readBytes(client: Turbopuffer, mount: string, path: string) {
  return runPlan(client, readBytesPlan(mountNamespace(mount), path));
}

export function mkdir(client: Turbopuffer, mount: string, path: string) {
  return runPlan(client, mkdirPlan(mountNamespace(mount), path));
}

export function putText(
  client: Turbopuffer,
  mount: string,
  path: string,
  text: string,
  options: { mime?: string } = {},
) {
  return runPlan(client, putTextPlan(mountNamespace(mount), path, text, options.mime));
}

export function putBytes(
  client: Turbopuffer,
  mount: string,
  path: string,
  data: Uint8Array,
  options: { mime?: string } = {},
) {
  return runPlan(client, putBytesPlan(mountNamespace(mount), path, data, options.mime));
}

export function rm(client: Turbopuffer, mount: string, path: string, recursive = false) {
  return runPlan(client, rmPlan(mountNamespace(mount), path, recursive));
}
